### PR TITLE
Add API for providing SNI AsyncMapping

### LINKED ISSUE #2156: an idea to support dynamic https like (caddy server)
## Motivation
i hope if we could support  tls lazy load certificates by domain
this might help us in renew the certificates (using letsencrypt or other)
support dynamic https like (caddy server)
see https://caddyserver.com/docs/quick-starts/https
## Desired solution
add function to load certificates by domain as key 
we could implment this function to renew or generate ssl certificate

### PR DESCRIPTION
**Pull request by violetagg**
Fixes #2156

### CODE REVIEW

**Comment by ashraf-revo** [comment]
@violetagg thanks , this looks more efficient

**Review by pderop — APPROVED** [review]
LGTM

File: reactor-netty-core/src/main/java/reactor/netty/tcp/SslProvider.java [hunk 1]
```diff
@@ -704,13 +734,15 @@ public final Builder closeNotifyReadTimeoutMillis(long closeNotifyReadTimeoutMil
 		@Override
 		public Builder addSniMapping(String domainName, Consumer<? super SslContextSpec> sslProviderBuilder) {
 			addInternal(domainName, sslProviderBuilder);
+			this.sniMappings = null;
```
Comment (simonbasle) [inline reactor-netty-core/src/main/java/reactor/netty/tcp/SslProvider.java:737]: should we mention in the javadoc that this method and its sisters are exclusive with the `AsyncMapping`-based one?

File: reactor-netty-core/src/main/java/reactor/netty/tcp/SslProvider.java [hunk 2]
```diff
@@ -719,6 +751,14 @@ public Builder setSniMappings(Map<String, Consumer<? super SslContextSpec>> conf
 			Objects.requireNonNull(confPerDomainName);
 			this.confPerDomainName.clear();
 			confPerDomainName.forEach(this::addInternal);
+			this.sniMappings = null;
+			return this;
+		}
+
+		@Override
+		public Builder setSniMappings(AsyncMapping<String, SslProvider> mappings) {
```
Comment (simonbasle) [inline reactor-netty-core/src/main/java/reactor/netty/tcp/SslProvider.java:759]: consider maybe renaming it slightly to indicate this is not entirely compatible with the legacy methods? something like `setSniAsyncMappings`?

**Review by simonbasle — COMMENTED** [review]
a couple interrogations about the way the configuration is presented to the user, but otherwise looks good.

File: reactor-netty-core/src/main/java/reactor/netty/tcp/SslProvider.java (hunk 1 above)
Comment (violetagg) [reply reactor-netty-core/src/main/java/reactor/netty/tcp/SslProvider.java:740]: fixed

File: reactor-netty-core/src/main/java/reactor/netty/tcp/SslProvider.java (hunk 2 above)
Comment (violetagg) [reply reactor-netty-core/src/main/java/reactor/netty/tcp/SslProvider.java]: fixed

File: reactor-netty-core/src/main/java/reactor/netty/tcp/SslProvider.java [hunk 3]
```diff
@@ -177,7 +177,8 @@ public interface Builder {
 		/**
 		 * Adds a mapping for the given domain name to an {@link SslProvider} builder.
 		 * If a mapping already exists, it will be overridden.
-		 * Note: This configuration is applicable only when configuring the server.
+		 * <p><strong>Note:</strong> This method is an alternative of {@link #setSniAsyncMappings(AsyncMapping)}.
```
[the remark this answers is not in the document]
Comment (violetagg) [reply reactor-netty-core/src/main/java/reactor/netty/tcp/SslProvider.java]: fixed

**Review by simonbasle — APPROVED** [review]
great  👍

**Comment by violetagg** [comment]
@pderop @simonbasle Thanks for the review!

**Pull request merged**